lib-cslam/src/duckietown_cslam/duckietownGraphBuilder/duckietown_graph_builder.py

```diff
@@ -4,234 +4,578 @@
+           associated to that node, outputs an integer that can be used as a
+           index for the node in g2o (the latter only handles integer indices
+           for the nodes).
+
+           Args:
+              id: ID in the format <node_type>_<node_id>
+              time_stamp: Timestamp.
+
+           Returns:
+              Input ID converted to an integer that can be used as an index by
+              g2o.
+        """
         [node_type, node_id] = id.split("_")
         a = 0
         b = int(node_id) % 1000
 
-        if(node_type == "duckie"):
+        if (node_type == "duckie"):
             a = 1
-        elif(node_type == "watchtower"):
+        elif (node_type == "watchtower"):
             a = 2
-        elif(node_type == "apriltag"):
+        elif (node_type == "apriltag"):
             a = 3
 
-        if(node_type in self.movable):
-            c = self.lists[node_type][node_id][time_stamp]
-            if(c >= 100000):
+        if (node_type in self.movable):
+            c = self.timestamp_local_indices[node_type][node_id][time_stamp]
+            if (c >= 100000):
                 print("overflow of the time_stamp list")
         else:
             c = 0
 
         result = a * 10**8 + b * 10**5 + c
         return result
 
-    def interpolate(self, old_time_stamp, new_time_stamp, node_type, node_id, measure):
-        vertexId = "%s_%s" % (node_type, node_id)
-        to_interpolate = {time_stamp: self.lists[node_type][node_id][time_stamp] for time_stamp in self.lists[node_type][node_id].keys() if (
-            time_stamp >= old_time_stamp and time_stamp <= new_time_stamp)}
-        # print("interpolating one odometry measure on %d nodes" %
-        #       len(to_interpolate))
-        sorted_time_stamps = sorted(to_interpolate.keys())
+    def retrointerpolate(self, time_stamp, node_type, node_id, vertex_id):
+        """Due to delays in the communication network, it might happen that a
+           message with timestamp time_stamp is received after a message with
+           timestamp time_stamp2, with time_stamp < time_stamp2. If this is the
+           case, this function inserts a vertex with timestamp time_stamp by
+           interpolating between the two timestamps - among those that have
+           already a vertex in the graph - immediately before and
+           immediately after time_stamp. This process is referred to as
+           retro-interpolation. Note: if time_stamp happens to follow
+           or precede in time all the timestamps with a vertex in the graph,
+           retro-interpolation cannot be performed.
 
+           Args:
+               time_stamp: Timestamp
+               node_type: Type of the node.
+               node_id: ID of the node.
+               vertex_id: ID in the format <node_type>_<node_id>.
+        """
+        # Find timestamps immediately before and immediately after the given
+        # timestamp.
+        time_stamp_before_is_set = False
```

Comment:
This is the part in `retrointerpolate` that I 'optimized' avoiding sorting the vector. We could even avoid having to check for each timestamp if `time_stamp_before_is_set`  (line 314, or `time_stamp_after_is_set`), but we would have to initialize `time_stamp_after` to either an existing timestamp that is larger than the current timestamp (`time_stamp`) or to a value that is for sure larger than the true `time_stamp_after` (something like INF). (Likewise for `time_stamp_before`). Both ways would either introduce some computational overhead (need to go through the vector again) or require to deal with possible overflows/knowning the maximum possible value of the timestamps, so I ended up with this.